### PR TITLE
feat: only allow dispatch for FrameConnector

### PR DESCRIPTION
**Pull request by Nebulis**
BREAKING CHANGE: removed props `methods` in FrameConnector component and provide only a dispatch method to the application

Got rid of `isActionOf` because it forces to use typesafe actions

### CODE REVIEW

File: src/components/frame/FrameConnector.tsx [hunk 1]
```diff
@@ -46,50 +35,44 @@ interface LegacyFrameConnectorProps extends BaseFrameConnectorProps {
...
+            toFrame.renderDocument(action.payload.document, action.payload.rawDocument);
+          } else if (action.type === "SELECT_TEMPLATE" && toFrame.selectTemplateTab) {
+            if (typeof action.payload === "number") {
+              toFrame.selectTemplateTab(action.payload);
+            } else if (action.meta.templates) {
+              const index = action.meta.templates.findIndex(template => template.id === action.payload);
+              toFrame.selectTemplateTab(index);
+            } else {
+              throw new Error(`Unable to handle ${action.type} when payload is a string`);
             }
```
Comment (yehjxraymond) [inline src/components/frame/FrameConnector.tsx:71]: The nested ifs are a little hard to read...

Comment (Nebulis) [reply src/components/frame/FrameConnector.tsx:71]: I need to add a comment on a specific case (number vs string) but except that suggestion?

File: src/components/frame/FrameConnector.tsx [hunk 2]
```diff
@@ -46,50 +35,44 @@ interface LegacyFrameConnectorProps extends BaseFrameConnectorProps {
...
-            print: () => {
-              if (toFrame.print) {
-                toFrame.print();
-                return true;
-              }
-              return false;
+        Object.assign((action: HostActions) => {
+          if (toFrame.dispatch) {
+            toFrame.dispatch(action);
+          } else if (action.type === "RENDER_DOCUMENT" && toFrame.renderDocument) {
```
Comment (yehjxraymond) [inline src/components/frame/FrameConnector.tsx:61]:
```suggestion
          if (dispatchExist) {
            toFrame.dispatch(action);
          } else {
            handleLegacyMethods(...);
          }
```

Thinking of something like this to better explain the difference

File: src/components/frame/HostConnector.tsx [hunk 3]
```diff
@@ -30,7 +29,7 @@ export const HostConnector: FunctionComponent<HostConnectorProps> = ({ dispatch,
         dispatch(renderDocument({ document, rawDocument }));
       },
       selectTemplateTab: (tabIndex: number) => {
-        dispatch(selectTemplate(tabIndex));
+        dispatch(selectTemplate(tabIndex, {}));
```
Comment (yehjxraymond) [inline src/components/frame/HostConnector.tsx:32]: why is there a `{}` here?

File: src/components/frame/FrameConnector.tsx (hunk 2 above)
Comment (Nebulis) [reply src/components/frame/FrameConnector.tsx:61]: I followed that pattern if/else but I kept the logic within the else block (no indirection to another function) + I added comments

File: src/components/frame/HostConnector.tsx (hunk 3 above)
Comment (Nebulis) [reply src/components/frame/HostConnector.tsx:32]: removed by keeping the templates internally

File: src/components/frame/host.actions.ts [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
+import { Document, Template } from "../../types";
 import { WrappedDocument } from "@govtechsg/open-attestation";
 
 export const renderDocument = createStandardAction("RENDER_DOCUMENT")<{
   document: Document;
   rawDocument?: WrappedDocument<Document>;
 }>();
-export const selectTemplate = createStandardAction("SELECT_TEMPLATE")<string | number>();
+
+export const selectTemplate = createStandardAction("SELECT_TEMPLATE")<string | number, { templates?: Template[] }>();
```
Comment (Nebulis) [inline src/components/frame/host.actions.ts]: THIS IS NOT NEEDED ANYMORE IMHO

Comment (Nebulis) [reply src/components/frame/host.actions.ts]: done

File: src/components/frame/useFrame.ts [hunk 5]
```diff
@@ -9,11 +9,8 @@ type Status = "DISCONNECTED" | "CONNECTING" | "CONNECTED";
 
 interface UseParentFrameProps {
   dispatch: HostActionsHandler;
-  methods?: LegacyHostActions;
 }
-export const useParentFrame = function(
-  props: UseParentFrameProps
-): [boolean, Partial<LegacyFrameActions> & { dispatch: FrameActionsHandler }] {
+export const useParentFrame = function(props: UseParentFrameProps): [boolean, { dispatch: FrameActionsHandler }] {
```
Comment (yehjxraymond) [inline src/components/frame/useFrame.ts:13]:
```suggestion
export const useParentFrame = function({dispatch}: UseParentFrameProps): [boolean, { dispatch: FrameActionsHandler }] {
```

Why not just destructure it?

File: src/components/frame/FrameConnector.tsx [hunk 6]
```diff
@@ -46,50 +36,64 @@ interface LegacyFrameConnectorProps extends BaseFrameConnectorProps {
...
+            // in that event we will map each action to it's legacy renderer function
+            // - "RENDER_DOCUMENT" must call toFrame.renderDocument
+            // - "SELECT_TEMPLATE" must call toFrame.selectTemplateTab
+            // - "PRINT" must call toFrame.print
+            if (action.type === "RENDER_DOCUMENT" && toFrame.renderDocument) {
+              toFrame.renderDocument(action.payload.document, action.payload.rawDocument);
+            } else if (action.type === "SELECT_TEMPLATE" && toFrame.selectTemplateTab) {
+              const index = templates.current.findIndex(template => template.id === action.payload);
+              if (index === -1) console.error(`Unable to find the index associated with the label ${action.payload}`);
+              toFrame.selectTemplateTab(index);
```
Comment (yehjxraymond) [inline src/components/frame/FrameConnector.tsx:91]: I thought you mentioned there is a case here where the action.payload could be a number also? Are we going to handle that?

Comment (Nebulis) [reply src/components/frame/FrameConnector.tsx:91]: I got rid of that in the latest commit
- application will use dispatch and fallback to legacy functions
- renderer will use dispatch only
- we will migrate our 2 apps to the latest renderer => no problem

File: src/components/frame/useFrame.ts (hunk 5 above)
Comment (Nebulis) [reply src/components/frame/useFrame.ts]: changed :)

**Pull request merged**